### PR TITLE
Feature: Handle symbol-to-proc wrappers (`&:`) that refer to a method created using delegation or method missing

### PR DESCRIPTION
**Pull request by marcrohloff**
## Summary

Improve handling of symbol-to-proc exposure blocks (`&:method_name`) for delegated and dynamically dispatched methods.

Previously grape-entity eagerly validated method arity for `&:method_name` blocks. This worked for ordinary methods, but produced incorrect behavior for methods implemented via `delegate` or `method_missing`, where Ruby may expose ambiguous or negative arity values.

This PR makes arity validation more conservative and Ruby-compatible:

* continue raising curated errors for ordinary methods with required arguments
* correctly allow methods with only optional args/splats/kwargs
* gracefully handle delegated and `method_missing`-backed methods
* fall back to Ruby's native runtime errors when introspection is unreliable

## Changes

* tighten symbol-to-proc wrapper detection
* add `positional_arity_for` helper for normalized arity handling
* support optional positional/keyword arguments correctly
* avoid false positives for delegation wrappers and dynamic dispatch
* add extensive spec coverage for delegation, `method_missing`, splats, kwargs, and edge cases

## Backward Compatibility

Public API is unchanged.

The only observable behavior change is that delegated or dynamic-dispatch methods may now raise Ruby-native errors instead of grape-entity-generated arity errors when the callable contract cannot be reliably introspected.

### CODE REVIEW

File: lib/grape_entity/entity.rb [hunk 1]
```diff
@@ -548,7 +548,8 @@ def ensure_block_arity!(block)
       end
 
       arity = object.method(origin_method_name).arity
-      return if arity.zero?
+      # functions defined using `delegate` or `method_missing` have an arity of -1
+      return if arity <= 0
```
Comment (numbata) [inline lib/grape_entity/entity.rb]: This is too permissive and reintroduces the bug class #389 was meant to prevent. Ruby's negative arities encode `-(required + 1)`, so `def foo(a, *rest)` has `arity -2` and `def foo(a, b, *rest)` has arity `-3` - both have required positional
args, but both now sail through this guard and blow up later at `instance_exec(object, &block)` with a "wrong number of arguments" raised from inside user code, instead of the friendly `ArgumentError` this method was designed to produce.

Worth noting: `def method_missing(method, ...)` itself has arity `-2`, so even the stated motivating case is partially mishandled by `<= 0`.

Suggested fix is match the actual intent ("no required positional args"):

```ruby
arity = object.method(origin_method_name).arity
required = arity >= 0 ? arity : -arity - 1
return if required.zero?
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,7 @@
 ### Next Release
 
+* [#406](https://github.com/ruby-grape/grape-entity/pull/406): Handle symbol-to-proc wrappers (`&:method_name`) where the method uses `delegate` or `method_missing` - [@marcrohloff](https://github.com/marcrohloff).
```
Comment (numbata) [inline CHANGELOG.md]: The entry is placed above the `#### Features` / `#### Fixes` headers rather than under one of them. Please move the line under `#### Fixes`.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -44,7 +44,7 @@ Implement your feature or bug fix.
 
 Ruby style is enforced with [Rubocop](https://github.com/bbatsov/rubocop), run `bundle exec rubocop` and fix any style issues highlighted.
 
-Make sure that `bundle exec rake` completes without errors.
+Make sure that `bundle exec rake` and `rubocop` completes without errors.
```
Comment (numbata) [inline CONTRIBUTING.md]:
```suggestion
Make sure that `bundle exec rake` and `rubocop` complete without errors.
```

Subject-verb agreement: two subjects -> complete. Also the line above already instructs users to run bundle exec rubocop, so this becomes a near-duplicate 🤷

File: spec/grape_entity/entity_spec.rb [hunk 4]
```diff
@@ -459,6 +484,28 @@ def raises_argument_error
             end
           end
 
+          context 'with block passed in via & that uses `missing_method`' do
+            specify do
+              subject.expose :using_missing, &:method_using_missing
+
+              object = SomeObject.new
+              expect(object.method(:method_using_missing).arity).to eq(-1)
```
Comment (numbata) [inline spec/grape_entity/entity_spec.rb]: The `value` assertion is the one that proves `grape-entity` handles it correctly. The arity line is testing Ruby.

```suggestion
```

File: spec/grape_entity/entity_spec.rb [hunk 5]
```diff
@@ -415,6 +423,23 @@ def method_with_multiple_args(_object, _options)
             def raises_argument_error
               raise ArgumentError, 'something different'
             end
+
+            def method_missing(method, ...)
```
Comment (numbata) [inline spec/grape_entity/entity_spec.rb:455]: Rather than adding `method_missing` directly to `SomeObject`, could you extract the new fixtures into subclasses? That way the shared class stays clean and the existing tests (especially the "undefined method" one) don't depend on your
respond_to_missing? being correct to pass.

Something like:

```ruby
SomeObjectWithMissing = Class.new(SomeObject) do
  def method_missing(method, ...)
    return 'missing-result' if method == :method_using_missing
    super
  end

  def respond_to_missing?(method, include_private = false)
    method == :method_using_missing || super
  end
end

SomeObjectWithDelegation = Class.new(SomeObject) do
  InnerDelegate = Class.new { def delegated_method = 'delegated-result' }
  delegate :delegated_method, to: :inner
  def inner = @inner ||= InnerDelegate.new
end
```

and specs kind of:

```ruby
context 'with &: referencing a method_missing-backed method' do
  specify do
    subject.expose :using_missing, &:method_using_missing

    object = SomeObjectWithMissing.new
    expect(subject.represent(object).value_for(:using_missing)).to eq('missing-result')
  end
end

context 'with &: referencing a delegated method' do
  specify do
    subject.expose :using_delegation, &:delegated_method

    object = SomeObjectWithDelegation.new
    expect(subject.represent(object).value_for(:using_delegation)).to eq('delegated-result')
  end
end
```
with  the regression case that pins the boundary:
```ruby
context 'with &: referencing a method with required args and a splat (arity -2)' do
  specify do
    subject.expose :x, &:method_with_required_and_splat

    object = SomeObjectWithMissing.new  # or any object with such a method
    expect do
      subject.represent(object).value_for(:x)
    end.to raise_error(ArgumentError, match(/method expects/))
  end
end
```

**Review by numbata — CHANGES_REQUESTED** [review]
@marcrohloff Thanks for the contribution! The motivation is solid. 👍 

Two things to address before merging: the `arity <= 0` guard is too wide (see inline comment), and the new `method_missing` in the shared fixture risks making the "undefined method" test fragile. Left suggestions on both.

File: spec/grape_entity/entity_spec.rb (hunk 5 above)
Comment (marcrohloff) [reply spec/grape_entity/entity_spec.rb:455]: @numbata 

I decided to add some more specs to test the negative cases. Something like the others tests. i..e
```
subject.expose :that_method_with_required_and_optional_args, &:method_with_required_and_optional_args
subject.expose :method_with_required_and_optional_args, as: :that_method_with_required_and_optional_args_again
```

Getting the value of the first `expose` gives the expected exception message. But the second results in `wrong number of arguments (given 0, expected 1)` because it is processed through the delegator

Do you think that the second message should match the first message?

Comment (numbata) [reply spec/grape_entity/entity_spec.rb:455]: Good question! My short subjective answer is: no, they don't need to match for this PR.

The `ensure_block_arity!` only guards the `&:symbol` path — the `as:` alias goes through different machinery and its error message is a separate concern. Aligning the two would be a worthwhile follow-up, but let's not expand scope here.

For the context: `as:` bypasses `ensure_block_arity!` not by skipping it directly, but because it never produces a `proc` at all, so the exposure type is `DelegatorExposure` (plain attribute delegation) rather than `BlockExposure`. Two completely different code paths.

File: spec/grape_entity/entity_spec.rb (hunk 4 above)
Comment (marcrohloff) [reply spec/grape_entity/entity_spec.rb]: It wasn't so much for testing as documenting an unexpected value. I removed it

File: CONTRIBUTING.md (hunk 3 above)
Comment (marcrohloff) [reply CONTRIBUTING.md]: You're right,m I reverted this change

File: spec/grape_entity/entity_spec.rb (hunk 5 above)
Comment (marcrohloff) [reply spec/grape_entity/entity_spec.rb:455]: I made the suggested changes and added several tests to verify that `ArgumentError` was raised when expected

File: lib/grape_entity/entity.rb [hunk 6]
```diff
@@ -547,11 +547,14 @@ def ensure_block_arity!(block)
...
+      # Ensure that the function does not require any positional args
+      # (functions defined using `delegate` or `method_missing` take an arg of `*rest`
       arity = object.method(origin_method_name).arity
-      return if arity.zero?
+      required_positional_arg_count = arity >= 0 ? arity : -arity - 1
+      return if required_positional_arg_count.zero?
 
       raise ArgumentError, <<~MSG
-        Cannot use `&:#{origin_method_name}` because that method expects #{arity} argument#{'s' if arity != 1}.
+        Cannot use `&:#{origin_method_name}` because that method expects #{required_positional_arg_count} #{'argument'.pluralize(required_positional_arg_count)}.
```
Comment (numbata) [inline lib/grape_entity/entity.rb]: The `.pluralize` is a ActiveSupport's method which requires `active_support/core_ext/string/inflections`.  It works today, but it depends on I18n config and inflection rules someone could override in a Rails app. How about `#{required_positional_arg_count == 1 ? 'argument' : 'arguments'}`? Zero magic, zero risk.

File: spec/grape_entity/entity_spec.rb [hunk 7]
```diff
@@ -408,6 +408,22 @@ def method_with_one_arg(_object)
...
+
+            def method_with_required_and_optional_args(_required_arg, _optional1 = 1, _optional2 = 2)
+              'result'
+            end
+
+            def method_with_optional_args_as_a_splat(*_optional_args)
+              'result'
+            end
+
+            def method_with_required_args_and_an_optional_splat(_required_arg, *_optional_argds)
```
Comment (numbata) [inline spec/grape_entity/entity_spec.rb]:
```suggestion
            def method_with_required_args_and_an_optional_splat(_required_arg, *_optional_args)
```

File: lib/grape_entity/entity.rb (hunk 6 above)
Comment (marcrohloff) [reply lib/grape_entity/entity.rb]: I reverted to the original syntax instead. I also fixed the other typo

**Comment by numbata** [comment]
Thanks for taking my comments into account, @marcrohloff! I added a few improvements on top of your commits before merging:

- Tightened the `Proc#to_s` regex to a strict anchored match to avoid false positives on _non-symbol-to-proc_ lambdas
- Fixed the arity math for variadic methods (`def foo(req, *rest)` was being miscounted)
- Added a bail-out for required keyword arguments so Ruby surfaces its native "missing keyword" error rather than a
misleading one from us
- Extracted `positional_arity_for` to keep the arity introspection logic self-contained and testable
- Expanded the spec coverage to include optional kwargs, splats, private methods, delegation-backed objects, and a canary spec for the `Proc#to_s` format the detection relies on

Great find, and thanks again for the PR! 👍

**Review by numbata — APPROVED** [review]

**Comment by rc-marc-rohloff** [comment]
@numbata Thanks for cleaning that up
